File: Algo/MoveZeroes.java
/* https://leetcode.com/problems/move-zeroes/ */
class MoveZeroes
{
	public static void main(String args[])
	{
		MoveZeroes solution = new MoveZeroes();
		// int[] nums = {0};
		// int[] nums = {1, 0};
		// int[] nums = {0, 1};
		// int[] nums = {0, 0, 0};
		// int[] nums = {1,0,1};
		// int[] nums = {1, 2, 3};
		int[] nums = {0, 1, 0, 3, 12};
		solution.moveZeroes(nums);

	}

	public void moveZeroes(int[] nums)
	{
			int ptr = 0;
			int i = 0;
			while(i<nums.length){
				if(nums[i]!=0){
					nums[ptr++] = nums[i];
				}
				i++;
			}
			while(ptr<nums.length){
				nums[ptr++] = 0;
			}
      for(int num: nums)
        System.out.print(num+"\t");
    }
}

File: Algo/python/MoveZeroes.py
''' https://leetcode.com/problems/move-zeroes/ '''

class Solution(object):
    def moveZeroes(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        zIndex = 0
        nzIndex = 0
        ln = len(nums)
        if ln < 2:
            return
        
        while zIndex < ln and nzIndex < ln:
            while zIndex < ln and nums[zIndex] != 0:
                zIndex += 1
        
            nzIndex = zIndex
            while nzIndex < ln and nums[nzIndex] == 0:
                nzIndex += 1
                
            #print zIndex, nzIndex, ln
            if zIndex < nzIndex and nzIndex < ln:
                nums[zIndex] = nums[nzIndex]
                nums[nzIndex] = 0
                zIndex += 1
                nzIndex += 1
            else:
                break
        
